### ISSUE TITLE
import-private-name does not trigger on external import of private modules

### ISSUE THREAD

**Issue by AiyionPrime** [pylint-dev/pylint#9160]
### Bug description

```python
Consider the following structure:


src/my_main_package/
├── __init__.py
├── accessor_broken1.py
├── accessor_broken2.py
└── my_package
    ├── __init__.py
    ├── _protected_sub_module.py
    └── valid_internal_access.py


`my_package._protected_sub_module` contains a class `SomeClass` like this:

```python
"""This submodule is protected and its public members should only be accessed by siblings in the same module."""


class SomeClass:
    """This class is public in a protected module and should be accessible by sibling files, like valid_access.py."""
```

Its siblings `valid_internal_access.py` and the `__init__.py` (both in `my_package`) are the only two files that should be allowed to import from it.

This what `valid_internal_access.py` could look like:

```python
"""This should work and could be used to expose SomeClass in the __init__.py"""
from _protected_submodule import SomeClass

SomeClass()
```

Consider the `__init__.py` in `my_package` does **not** import the protected module.
And therefore no expose it to other external classes.

There are then two broken accessor examples, which both should not be able to import the protected class without warnings:

```python accessor_broken1.py
"""This is an external accessor, which should not access _protected_submodule like this."""
from voraus_no_core.my_package import _protected_submodule

t2 = _protected_submodule.SomeClass()
```

and variant two:

```python accessor_broken2.py
"""This should not either."""
from voraus_no_core.my_package._protected_submodule import SomeClass

t = SomeClass()
```
```


### Configuration

```ini
[MAIN]
load-plugins= """
    pylint.extensions.private_import
    """
```

### Command used

```shell
pylint .
```

### Pylint output

```python
-------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 8.57/10, +1.43)
```

### Expected behavior

From the description of **C2701** I would have expected this to result in two pylint errors.
Am I missing something? Or is this a bug?

> Used when a **private module** or object prefixed with _ is imported.

__https://pylint.pycqa.org/en/stable/user_guide/messages/convention/import-private-name.html__

### Pylint version

```shell
pylint 2.17.4
astroid 2.15.6
Python 3.9.18 (main, Oct 17 2023, 12:33:47) 
[GCC 13.2.1 20230801]
```

### OS / Environment

Archlinux (should not be relevant)

**Comment by mbyrnepr2** [comment]
Interesting one @AiyionPrime. I looked into it and I believe this is expected behaviour to not emit the `import-private-name` for modules within the same codebase. Relevant logic [here](https://github.com/pylint-dev/pylint/blob/main/pylint/extensions/private_import.py#L61-L63).

In the [documentation](https://pylint.pycqa.org/en/stable/user_guide/messages/convention/import-private-name.html) I think the key point is indicated at the end of the page since it mentions dependencies:
_Using private imports expose you to unexpected breaking changes for any version bump of your dependencies, even in patch versions._

**Comment by mbyrnepr2** [comment]
Looks like there is a false positive, however, in the case of `valid_internal_access.py` having this code: `from _protected_submodule import SomeClass` since the common base directory is not found in this case.

**Comment by AiyionPrime** [comment]
So for what I've got in mind I need to propose a new feature like `import-private-module`?

@areveny Is it correct that the external import of private modules in the same project/codebase is in fact out of scope of your extension?

**Comment by mbyrnepr2** [comment]
Please do @AiyionPrime. We can keep it separate from this issue which can be retitled to focus on the false positive.
